File: Models/DatabaseWorker.cs
/* File: DatabaseWorker.cs
 * Author: Glenn Sutherland, ChatGPT Codex
 * Description: This manages the sql commands and database for the software. 
*/
using Spotify_Playlist_Manager.Models;
using System;
using System.Collections.Generic;
using Microsoft.Data.Sqlite;
namespace Spotify_Playlist_Manager.Models
{
    
    public static class DatabaseWorker
    {
        private static string _dbPath = Variables.DatabasePath;

        public static void Init()
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();

            cmd.CommandText = """
                                  -- Settings table
                                  CREATE TABLE IF NOT EXISTS Settings (
                                      Key TEXT PRIMARY KEY,
                                      Value TEXT NOT NULL
                                  );

                                  -- Playlists table
                                  CREATE TABLE IF NOT EXISTS Playlists (
                                      Id TEXT PRIMARY KEY,               -- Spotify playlist ID
                                      Name TEXT,
                                      ImageURL TEXT,
                                      Description TEXT,
                                      SnapshotID TEXT,
                                      TrackIDs TEXT                      -- 'id1;;id2;;id3'
                                  );

                                  -- Albums table
                                  CREATE TABLE IF NOT EXISTS Albums (
                                      Id TEXT PRIMARY KEY,               -- Spotify album ID
                                      Name TEXT,
                                      ImageURL TEXT,
                                      ArtistIDs TEXT                     -- 'id1;;id2;;id3'
                                  );

                                  -- Tracks table (Spotify ID is anchor; SongID is internal secondary)
                                  CREATE TABLE IF NOT EXISTS Tracks (
                                      Id TEXT PRIMARY KEY,               -- Spotify track ID
                                      SongID TEXT,                       -- Carillon internal ID (CIID)
                                      Name TEXT,
                                      AlbumId TEXT,
                                      ArtistIds TEXT,
                                      DiscNumber INTEGER,
                                      DurationMs INTEGER,
                                      Explicit INTEGER,                  -- 0 = false, 1 = true
                                      PreviewUrl TEXT,
                                      TrackNumber INTEGER
                                  );

                                  -- Artists table
                                  CREATE TABLE IF NOT EXISTS Artists (
                                      Id TEXT PRIMARY KEY,               -- Spotify artist ID
                                      Name TEXT,
                                      ImageURL TEXT,
                                      Genres TEXT                        -- fixed typo: "Generes" → "Genres"
                                  );

                                  -- Similar table
                                  CREATE TABLE IF NOT EXISTS Similar (
                                      SongID TEXT,                       -- Spotify similar ID
                                      SongID2 TEXT                       -- removed trailing comma
                                  );
                              """;


            cmd.ExecuteNonQuery();
        }


        public static void SetSetting(string key, string value)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();
            using var cmd = conn.CreateCommand();
            cmd.CommandText = "INSERT OR REPLACE INTO Settings (Key, Value) VALUES ($key, $value);";
            cmd.Parameters.AddWithValue("$key", key);
            cmd.Parameters.AddWithValue("$value", value);
            cmd.ExecuteNonQuery();
        }

        public static string? GetSetting(string key)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();
            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Value FROM Settings WHERE Key = $key;";
            cmd.Parameters.AddWithValue("$key", key);
            return cmd.ExecuteScalar() as string;
        }

        public static IEnumerable<(string key, string value)> GetAllSettings()
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Key, Value FROM Settings;";

            using var reader = cmd.ExecuteReader();

            // Read each row and yield it as a tuple
            while (reader.Read())
            {
                string key = reader.GetString(0);   // column index 0 → Key
                string value = reader.GetString(1); // column index 1 → Value
                yield return (key, value);
            }
        }

        public static void SetPlaylist(Variables.PlayList playlist)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = @"INSERT OR REPLACE INTO Playlists (Id, Name, ImageURL, Description, SnapshotID, TrackIDs)
                                VALUES ($id, $name, $imageUrl, $description, $snapshotId, $trackIds);";

            cmd.Parameters.AddWithValue("$id", playlist.Id ?? string.Empty);
            cmd.Parameters.AddWithValue("$name", playlist.Name ?? string.Empty);
            cmd.Parameters.AddWithValue("$imageUrl", playlist.ImageURL ?? string.Empty);
            cmd.Parameters.AddWithValue("$description", playlist.Description ?? string.Empty);
            cmd.Parameters.AddWithValue("$snapshotId", playlist.SnapshotID ?? string.Empty);
            cmd.Parameters.AddWithValue("$trackIds", playlist.TrackIDs ?? string.Empty);

            cmd.ExecuteNonQuery();
        }

        public static Variables.PlayList? GetPlaylist(string id)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Id, Name, ImageURL, Description, SnapshotID, TrackIDs FROM Playlists WHERE Id = $id LIMIT 1;";
            cmd.Parameters.AddWithValue("$id", id);

            using var reader = cmd.ExecuteReader();

            if (reader.Read())
            {
                return new Variables.PlayList
                {
                    Id = reader["Id"]?.ToString() ?? string.Empty,
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    ImageURL = reader["ImageURL"]?.ToString() ?? string.Empty,
                    Description = reader["Description"]?.ToString() ?? string.Empty,
                    SnapshotID = reader["SnapshotID"]?.ToString() ?? string.Empty,
                    TrackIDs = reader["TrackIDs"]?.ToString() ?? string.Empty
                };
            }

            return null;
        }

        public static IEnumerable<Variables.PlayList> GetAllPlaylists()
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Id, Name, ImageURL, Description, SnapshotID, TrackIDs FROM Playlists;";

            using var reader = cmd.ExecuteReader();

            while (reader.Read())
            {
                yield return new Variables.PlayList
                {
                    Id = reader["Id"]?.ToString() ?? string.Empty,
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    ImageURL = reader["ImageURL"]?.ToString() ?? string.Empty,
                    Description = reader["Description"]?.ToString() ?? string.Empty,
                    SnapshotID = reader["SnapshotID"]?.ToString() ?? string.Empty,
                    TrackIDs = reader["TrackIDs"]?.ToString() ?? string.Empty
                };
            }
        }

        public static void SetAlbum(Variables.Album album)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = @"INSERT OR REPLACE INTO Albums (Id, Name, ImageURL, ArtistIDs)
                                VALUES ($id, $name, $imageUrl, $artistIds);";

            cmd.Parameters.AddWithValue("$id", album.Id ?? string.Empty);
            cmd.Parameters.AddWithValue("$name", album.Name ?? string.Empty);
            cmd.Parameters.AddWithValue("$imageUrl", album.ImageURL ?? string.Empty);
            cmd.Parameters.AddWithValue("$artistIds", album.ArtistIDs ?? string.Empty);

            cmd.ExecuteNonQuery();
        }

        public static Variables.Album? GetAlbum(string id)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Id, Name, ImageURL, ArtistIDs FROM Albums WHERE Id = $id LIMIT 1;";
            cmd.Parameters.AddWithValue("$id", id);

            using var reader = cmd.ExecuteReader();

            if (reader.Read())
            {
                var albumId = reader["Id"]?.ToString() ?? string.Empty;

                return new Variables.Album
                {
                    Id = albumId,
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    ImageURL = reader["ImageURL"]?.ToString() ?? string.Empty,
                    ArtistIDs = reader["ArtistIDs"]?.ToString() ?? string.Empty,
                    TrackIDs = GetAlbumTrackIds(conn, albumId)
                };
            }

            return null;
        }

        public static IEnumerable<Variables.Album> GetAllAlbums()
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Id, Name, ImageURL, ArtistIDs FROM Albums;";

            using var reader = cmd.ExecuteReader();

            while (reader.Read())
            {
                var albumId = reader["Id"]?.ToString() ?? string.Empty;
                yield return new Variables.Album
                {
                    Id = albumId,
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    ImageURL = reader["ImageURL"]?.ToString() ?? string.Empty,
                    ArtistIDs = reader["ArtistIDs"]?.ToString() ?? string.Empty,
                    TrackIDs = GetAlbumTrackIds(conn, albumId)
                };
            }
        }

        private static string GetAlbumTrackIds(SqliteConnection connection, string albumId)
        {
            using var trackCmd = connection.CreateCommand();
            trackCmd.CommandText = "SELECT Id FROM Tracks WHERE AlbumId = $albumId;";
            trackCmd.Parameters.AddWithValue("$albumId", albumId);

            using var trackReader = trackCmd.ExecuteReader();
            List<string> trackIds = new();

            while (trackReader.Read())
            {
                var trackId = trackReader["Id"]?.ToString();
                if (!string.IsNullOrEmpty(trackId))
                {
                    trackIds.Add(trackId);
                }
            }

            return string.Join(Variables.Seperator, trackIds);
        }

        public static void SetTrack(Variables.Track track)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = @"INSERT OR REPLACE INTO Tracks (Id, SongID, Name, AlbumId, ArtistIds, DiscNumber, DurationMs, Explicit, PreviewUrl, TrackNumber)
                                VALUES ($id, $songId, $name, $albumId, $artistIds, $discNumber, $durationMs, $explicit, $previewUrl, $trackNumber);";

            cmd.Parameters.AddWithValue("$id", track.Id ?? string.Empty);
            cmd.Parameters.AddWithValue("$songId", track.SongID ?? string.Empty);
            cmd.Parameters.AddWithValue("$name", track.Name ?? string.Empty);
            cmd.Parameters.AddWithValue("$albumId", track.AlbumId ?? string.Empty);
            cmd.Parameters.AddWithValue("$artistIds", track.ArtistIds ?? string.Empty);
            cmd.Parameters.AddWithValue("$discNumber", track.DiscNumber);
            cmd.Parameters.AddWithValue("$durationMs", track.DurationMs);
            cmd.Parameters.AddWithValue("$explicit", track.Explicit ? 1 : 0);
            cmd.Parameters.AddWithValue("$previewUrl", track.PreviewUrl ?? string.Empty);
            cmd.Parameters.AddWithValue("$trackNumber", track.TrackNumber);

            cmd.ExecuteNonQuery();
        }

        public static Variables.Track GetTrack(string id)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();

            // Check if we're dealing with a Carillon internal SongID or a Spotify ID
            if (id.StartsWith(Variables.Identifier)) // e.g., "CIID"
            {
                cmd.CommandText = "SELECT * FROM Tracks WHERE SongID = @id LIMIT 1;";
            }
            else
            {
                cmd.CommandText = "SELECT * FROM Tracks WHERE Id = @id LIMIT 1;";
            }

            cmd.Parameters.AddWithValue("@id", id);

            using var reader = cmd.ExecuteReader();

            if (reader.Read())
            {
                // Safely convert columns to your Track object
                 return new Variables.Track() {
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    Id = reader["Id"]?.ToString() ?? string.Empty,
                    AlbumId = reader["AlbumId"]?.ToString() ?? string.Empty,
                    ArtistIds = reader["ArtistIds"]?.ToString() ?? string.Empty,
                    DiscNumber = reader["DiscNumber"] is DBNull ? 0 : Convert.ToInt32(reader["DiscNumber"]),
                    DurationMs = reader["DurationMs"] is DBNull ? 0 : Convert.ToInt32(reader["DurationMs"]),
                    Explicit = reader["Explicit"] is DBNull ? false : Convert.ToInt32(reader["Explicit"]) == 1,
                    PreviewUrl = reader["PreviewUrl"]?.ToString() ?? string.Empty,
                    TrackNumber = reader["TrackNumber"] is DBNull ? 0 : Convert.ToInt32(reader["TrackNumber"]),
                    SongID = reader["SongID"]?.ToString() ?? string.Empty
                };
            }

            return null; // No result found
        }

        public static IEnumerable<Variables.Track> GetAllTracks()
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT * FROM Tracks;";

            using var reader = cmd.ExecuteReader();

            while (reader.Read())
            {
                yield return new Variables.Track() {
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    Id = reader["Id"]?.ToString() ?? string.Empty,
                    AlbumId = reader["AlbumId"]?.ToString() ?? string.Empty,
                    ArtistIds = reader["ArtistIds"]?.ToString() ?? string.Empty,
                    DiscNumber = reader["DiscNumber"] is DBNull ? 0 : Convert.ToInt32(reader["DiscNumber"]),
                    DurationMs = reader["DurationMs"] is DBNull ? 0 : Convert.ToInt32(reader["DurationMs"]),
                    Explicit = reader["Explicit"] is DBNull ? false : Convert.ToInt32(reader["Explicit"]) == 1,
                    PreviewUrl = reader["PreviewUrl"]?.ToString() ?? string.Empty,
                    TrackNumber = reader["TrackNumber"] is DBNull ? 0 : Convert.ToInt32(reader["TrackNumber"]),
                    SongID = reader["SongID"]?.ToString() ?? string.Empty
                };
            }
        }

        public static void SetArtist(Variables.Artist artist)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = @"INSERT OR REPLACE INTO Artists (Id, Name, ImageURL, Genres)
                                VALUES ($id, $name, $imageUrl, $genres);";

            cmd.Parameters.AddWithValue("$id", artist.Id ?? string.Empty);
            cmd.Parameters.AddWithValue("$name", artist.Name ?? string.Empty);
            cmd.Parameters.AddWithValue("$imageUrl", artist.ImageURL ?? string.Empty);
            cmd.Parameters.AddWithValue("$genres", artist.Genres ?? string.Empty);

            cmd.ExecuteNonQuery();
        }

        public static Variables.Artist? GetArtist(string id)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Id, Name, ImageURL, Genres FROM Artists WHERE Id = $id LIMIT 1;";
            cmd.Parameters.AddWithValue("$id", id);

            using var reader = cmd.ExecuteReader();

            if (reader.Read())
            {
                return new Variables.Artist
                {
                    Id = reader["Id"]?.ToString() ?? string.Empty,
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    ImageURL = reader["ImageURL"]?.ToString() ?? string.Empty,
                    Genres = reader["Genres"]?.ToString() ?? string.Empty
                };
            }

            return null;
        }

        public static IEnumerable<Variables.Artist> GetAllArtists()
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT Id, Name, ImageURL, Genres FROM Artists;";

            using var reader = cmd.ExecuteReader();

            while (reader.Read())
            {
                yield return new Variables.Artist
                {
                    Id = reader["Id"]?.ToString() ?? string.Empty,
                    Name = reader["Name"]?.ToString() ?? string.Empty,
                    ImageURL = reader["ImageURL"]?.ToString() ?? string.Empty,
                    Genres = reader["Genres"]?.ToString() ?? string.Empty
                };
            }
        }

        public static void SetSimilar(string songId, string songId2)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = @"INSERT INTO Similar (SongID, SongID2) VALUES ($songId, $songId2);";

            cmd.Parameters.AddWithValue("$songId", songId);
            cmd.Parameters.AddWithValue("$songId2", songId2);

            cmd.ExecuteNonQuery();
        }

        public static (string SongId, string SongId2)? GetSimilar(string songId, string songId2)
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT SongID, SongID2 FROM Similar WHERE SongID = $songId AND SongID2 = $songId2 LIMIT 1;";
            cmd.Parameters.AddWithValue("$songId", songId);
            cmd.Parameters.AddWithValue("$songId2", songId2);

            using var reader = cmd.ExecuteReader();

            if (reader.Read())
            {
                return (
                    reader["SongID"]?.ToString() ?? string.Empty,
                    reader["SongID2"]?.ToString() ?? string.Empty
                );
            }

            return null;
        }

        public static IEnumerable<(string SongId, string SongId2)> GetAllSimilar()
        {
            using var conn = new SqliteConnection($"Data Source={_dbPath}");
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT SongID, SongID2 FROM Similar;";

            using var reader = cmd.ExecuteReader();

            while (reader.Read())
            {
                yield return (
                    reader["SongID"]?.ToString() ?? string.Empty,
                    reader["SongID2"]?.ToString() ?? string.Empty
                );
            }
        }


    }
}

File: carillon/database_worker.py
"""Database worker for sharing the existing Carillon SQLite database."""

from __future__ import annotations

__author__ = "ChatGPT Codex"

import os
import random
import sqlite3
import string
from dataclasses import dataclass
from itertools import islice
from pathlib import Path
from typing import Optional, TYPE_CHECKING

from appdirs import user_data_dir

if TYPE_CHECKING:
    from carillon.spotify_worker import SpotifyWorker


@dataclass
class DatabaseConfig:
    """Configuration for locating the shared SQLite database."""

    app_name: str = "SpotifyPlaylistManager"
    app_author: str = ""
    db_filename: str = "data.db"


class DatabaseWorker:
    """Thin wrapper around the shared SQLite database."""

    def __init__(self, config: DatabaseConfig | None = None) -> None:
        self._config = config or DatabaseConfig()
        self._connection: Optional[sqlite3.Connection] = None
        self._db_path = self._resolve_db_path()

    @property
    def db_path(self) -> Path:
        return self._db_path

    def _resolve_db_path(self) -> Path:
        base_dir = Path(user_data_dir(self._config.app_name, self._config.app_author))
        base_dir.mkdir(parents=True, exist_ok=True)
        return base_dir / self._config.db_filename

    def init(self) -> None:
        """Connect to the database and ensure the Settings table exists."""
        if self._connection is None:
            self._connection = sqlite3.connect(self._db_path)
            self._connection.execute("PRAGMA journal_mode=WAL;")

        self._connection.execute(
            """
            CREATE TABLE IF NOT EXISTS Settings (
                Key TEXT PRIMARY KEY,
                Value TEXT NOT NULL
            );
            """
        )
        self._connection.commit()

    def get_setting(self, key: str) -> Optional[str]:
        if self._connection is None:
            raise RuntimeError("DatabaseWorker.init must be called before get_setting.")
        cursor = self._connection.execute(
            "SELECT Value FROM Settings WHERE Key = ?;",
            (key,),
        )
        row = cursor.fetchone()
        return row[0] if row else None

    def set_setting(self, key: str, value: str) -> None:
        if self._connection is None:
            raise RuntimeError("DatabaseWorker.init must be called before set_setting.")
        self._connection.execute(
            "INSERT OR REPLACE INTO Settings (Key, Value) VALUES (?, ?);",
            (key, value),
        )
        self._connection.commit()

    def close(self) -> None:
        if self._connection is not None:
            self._connection.close()
            self._connection = None

    def sync_from_spotify(self, spotify: "SpotifyWorker") -> None:
        """
        Syncs local database with Spotify data before sorting begins.
        Ensures local 'sorted' status is up to date.
        """
        if self._connection is None:
            self._connection = sqlite3.connect(self._db_path)
        self._connection.row_factory = sqlite3.Row

        separator = ";;"

        def ensure_schema() -> None:
            self._connection.executescript(
                """
                CREATE TABLE IF NOT EXISTS Settings (
                    Key TEXT PRIMARY KEY,
                    Value TEXT NOT NULL
                );

                CREATE TABLE IF NOT EXISTS Playlists (
                    Id TEXT PRIMARY KEY,
                    Name TEXT,
                    ImageURL TEXT,
                    ImagePath TEXT,
                    Description TEXT,
                    SnapshotID TEXT,
                    TrackIDs TEXT
                );

                CREATE TABLE IF NOT EXISTS Albums (
                    Id TEXT PRIMARY KEY,
                    Name TEXT,
                    ImageURL TEXT,
                    ImagePath TEXT,
                    ArtistIDs TEXT
                );

                CREATE TABLE IF NOT EXISTS Tracks (
                    Id TEXT PRIMARY KEY,
                    SongID TEXT,
                    Name TEXT,
                    AlbumId TEXT,
                    ArtistIds TEXT,
                    DiscNumber INTEGER,
                    DurationMs INTEGER,
                    Explicit INTEGER,
                    PreviewUrl TEXT,
                    TrackNumber INTEGER
                );

                CREATE TABLE IF NOT EXISTS Artists (
                    Id TEXT PRIMARY KEY,
                    Name TEXT,
                    ImageURL TEXT,
                    ImagePath TEXT,
                    Genres TEXT
                );
                """
            )
            self._connection.execute("PRAGMA journal_mode=WAL;")
            self._connection.commit()

        def make_song_id(track_type: str = "SNG", length: int = 30) -> str:
            allowed_chars = string.ascii_uppercase + string.digits
            random_suffix = "".join(random.choice(allowed_chars) for _ in range(length))
            return f"CIID___{track_type}___{random_suffix}"

        def chunked(iterable, size: int):
            iterator = iter(iterable)
            while True:
                chunk = list(islice(iterator, size))
                if not chunk:
                    break
                yield chunk

        def fetch_playlist_track_ids(playlist_id: str) -> list[str]:
            track_ids: list[str] = []
            offset = 0
            while True:
                response = spotify.sp.playlist_items(
                    playlist_id,
                    limit=100,
                    offset=offset,
                    fields="items.track.id,next",
                )
                items = response.get("items", [])
                if not items:
                    break
                for item in items:
                    track = item.get("track") or {}
                    track_id = track.get("id")
                    if track_id:
                        track_ids.append(track_id)
                if response.get("next") is None:
                    break
                offset += 100
            return track_ids

        def ensure_track_placeholder(track_id: str) -> None:
            existing = self._connection.execute(
                "SELECT SongID FROM Tracks WHERE Id = ?;",
                (track_id,),
            ).fetchone()
            if existing:
                return
            self._connection.execute(
                """
                INSERT OR REPLACE INTO Tracks
                    (Id, SongID, Name, AlbumId, ArtistIds, DiscNumber, DurationMs, Explicit, PreviewUrl, TrackNumber)
                VALUES (?, ?, '', '', '', 0, 0, 0, '', 0);
                """,
                (track_id, make_song_id()),
            )

        def ensure_album_placeholder(album_id: str) -> None:
            existing = self._connection.execute(
                "SELECT Id FROM Albums WHERE Id = ?;",
                (album_id,),
            ).fetchone()
            if existing:
                return
            self._connection.execute(
                "INSERT OR REPLACE INTO Albums (Id, Name, ImageURL, ImagePath, ArtistIDs) VALUES (?, '', '', '', '');",
                (album_id,),
            )

        def ensure_artist_placeholder(artist_id: str) -> None:
            existing = self._connection.execute(
                "SELECT Id FROM Artists WHERE Id = ?;",
                (artist_id,),
            ).fetchone()
            if existing:
                return
            self._connection.execute(
                "INSERT OR REPLACE INTO Artists (Id, Name, ImageURL, ImagePath, Genres) VALUES (?, '', '', '', '');",
                (artist_id,),
            )

        ensure_schema()

        print("\n[Sync] Updating local database from Spotify...")

        # Sync playlists and their tracks
        print("[Sync] Playlists...")
        playlist_offset = 0
        playlists: list[dict] = []
        while True:
            response = spotify.sp.current_user_playlists(limit=50, offset=playlist_offset)
            items = response.get("items", [])
            if not items:
                break
            playlists.extend(items)
            if response.get("next") is None:
                break
            playlist_offset += 50

        playlist_snapshots = {
            row["Id"]: row["SnapshotID"]
            for row in self._connection.execute("SELECT Id, SnapshotID FROM Playlists;").fetchall()
        }

        for playlist in playlists:
            playlist_id = playlist["id"]
            details = spotify.sp.playlist(
                playlist_id,
                fields="name,images,description,snapshot_id",
            )
            snapshot_id = details.get("snapshot_id", "")
            if playlist_snapshots.get(playlist_id) == snapshot_id:
                continue
            image_url = ""
            images = details.get("images") or []
            if images:
                image_url = images[0].get("url", "")
            track_ids = fetch_playlist_track_ids(playlist_id)
            self._connection.execute(
                """
                INSERT OR REPLACE INTO Playlists
                    (Id, Name, ImageURL, ImagePath, Description, SnapshotID, TrackIDs)
                VALUES (?, ?, ?, '', ?, ?, ?);
                """,
                (
                    playlist_id,
                    details.get("name", ""),
                    image_url,
                    details.get("description", ""),
                    snapshot_id,
                    separator.join(track_ids),
                ),
            )
            for track_id in track_ids:
                ensure_track_placeholder(track_id)

        # Sync albums and their tracks
        print("[Sync] Albums...")
        album_offset = 0
        album_ids: list[str] = []
        while True:
            response = spotify.sp.current_user_saved_albums(limit=50, offset=album_offset)
            items = response.get("items", [])
            if not items:
                break
            for item in items:
                album = item.get("album") or {}
                album_id = album.get("id")
                if album_id:
                    album_ids.append(album_id)
            if response.get("next") is None:
                break
            album_offset += 50

        existing_albums = {
            row["Id"] for row in self._connection.execute("SELECT Id FROM Albums;").fetchall()
        }
        new_album_ids = [album_id for album_id in album_ids if album_id not in existing_albums]

        for album_batch in chunked(new_album_ids, 20):
            album_details = spotify.sp.albums(album_batch).get("albums", [])
            for album in album_details:
                if not album:
                    continue
                album_id = album.get("id")
                if not album_id:
                    continue
                artist_ids = [artist.get("id") for artist in album.get("artists", []) if artist.get("id")]
                track_ids: list[str] = []
                track_offset = 0
                while True:
                    tracks_resp = spotify.sp.album_tracks(album_id, limit=50, offset=track_offset)
                    track_items = tracks_resp.get("items", [])
                    if not track_items:
                        break
                    for track in track_items:
                        track_id = track.get("id")
                        if track_id:
                            track_ids.append(track_id)
                    if tracks_resp.get("next") is None:
                        break
                    track_offset += 50

                image_url = ""
                images = album.get("images") or []
                if images:
                    image_url = images[0].get("url", "")

                self._connection.execute(
                    """
                    INSERT OR REPLACE INTO Albums (Id, Name, ImageURL, ImagePath, ArtistIDs)
                    VALUES (?, ?, ?, '', ?);
                    """,
                    (
                        album_id,
                        album.get("name", ""),
                        image_url,
                        separator.join(artist_ids),
                    ),
                )
                for track_id in track_ids:
                    ensure_track_placeholder(track_id)
                for artist_id in artist_ids:
                    ensure_artist_placeholder(artist_id)

        # Sync liked songs
        print("[Sync] Liked songs...")
        liked_offset = 0
        while True:
            response = spotify.sp.current_user_saved_tracks(limit=50, offset=liked_offset)
            items = response.get("items", [])
            if not items:
                break
            for item in items:
                track = item.get("track") or {}
                track_id = track.get("id")
                if track_id:
                    ensure_track_placeholder(track_id)
            if response.get("next") is None:
                break
            liked_offset += 50

        # Sync track metadata
        print("[Sync] Track metadata...")
        track_rows = self._connection.execute("SELECT * FROM Tracks;").fetchall()
        missing_track_ids = []
        track_row_map = {}
        for row in track_rows:
            track_row_map[row["Id"]] = row
            if not row["Name"] or not row["AlbumId"] or not row["ArtistIds"] or not row["SongID"]:
                missing_track_ids.append(row["Id"])
                continue
            if row["DurationMs"] <= 0 or row["DiscNumber"] <= 0 or row["TrackNumber"] <= 0:
                missing_track_ids.append(row["Id"])

        for track_batch in chunked(missing_track_ids, 50):
            track_details = spotify.sp.tracks(track_batch).get("tracks", [])
            for track in track_details:
                if not track:
                    continue
                track_id = track.get("id")
                if not track_id:
                    continue
                row = track_row_map.get(track_id)
                song_id = row["SongID"] if row and row["SongID"] else make_song_id()
                artist_ids = [artist.get("id") for artist in track.get("artists", []) if artist.get("id")]
                album = track.get("album") or {}
                album_id = album.get("id", "")
                self._connection.execute(
                    """
                    INSERT OR REPLACE INTO Tracks
                        (Id, SongID, Name, AlbumId, ArtistIds, DiscNumber, DurationMs, Explicit, PreviewUrl, TrackNumber)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
                    """,
                    (
                        track_id,
                        song_id,
                        track.get("name", ""),
                        album_id,
                        separator.join(artist_ids),
                        track.get("disc_number") or 0,
                        track.get("duration_ms") or 0,
                        1 if track.get("explicit") else 0,
                        track.get("preview_url") or "",
                        track.get("track_number") or 0,
                    ),
                )
                if album_id:
                    ensure_album_placeholder(album_id)
                for artist_id in artist_ids:
                    ensure_artist_placeholder(artist_id)

        # Sync album metadata
        print("[Sync] Album metadata...")
        album_rows = self._connection.execute("SELECT * FROM Albums;").fetchall()
        missing_album_ids = [
            row["Id"]
            for row in album_rows
            if not row["Name"] or not row["ArtistIDs"]
        ]

        for album_batch in chunked(missing_album_ids, 20):
            album_details = spotify.sp.albums(album_batch).get("albums", [])
            for album in album_details:
                if not album:
                    continue
                album_id = album.get("id")
                if not album_id:
                    continue
                artist_ids = [artist.get("id") for artist in album.get("artists", []) if artist.get("id")]
                image_url = ""
                images = album.get("images") or []
                if images:
                    image_url = images[0].get("url", "")
                self._connection.execute(
                    """
                    INSERT OR REPLACE INTO Albums (Id, Name, ImageURL, ImagePath, ArtistIDs)
                    VALUES (?, ?, ?, '', ?);
                    """,
                    (
                        album_id,
                        album.get("name", ""),
                        image_url,
                        separator.join(artist_ids),
                    ),
                )
                for artist_id in artist_ids:
                    ensure_artist_placeholder(artist_id)

        # Sync artist metadata
        print("[Sync] Artist metadata...")
        artist_rows = self._connection.execute("SELECT * FROM Artists;").fetchall()
        missing_artist_ids = [
            row["Id"]
            for row in artist_rows
            if not row["Name"]
        ]

        for artist_batch in chunked(missing_artist_ids, 50):
            artist_details = spotify.sp.artists(artist_batch).get("artists", [])
            for artist in artist_details:
                if not artist:
                    continue
                artist_id = artist.get("id")
                if not artist_id:
                    continue
                images = artist.get("images") or []
                image_url = images[0].get("url", "") if images else ""
                self._connection.execute(
                    """
                    INSERT OR REPLACE INTO Artists (Id, Name, ImageURL, ImagePath, Genres)
                    VALUES (?, ?, ?, '', ?);
                    """,
                    (
                        artist_id,
                        artist.get("name", ""),
                        image_url,
                        ", ".join(artist.get("genres", []) or []),
                    ),
                )

        self._connection.commit()

        print("[Sync] Complete.")

    def __enter__(self) -> "DatabaseWorker":
        self.init()
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.close()
